proof/refine/ARM/Ipc_R.thy

```diff
@@ -2565,108 +3045,126 @@ lemma send_signal_corres:
+          apply (subgoal_tac "live (Notification ntfn)")
+           apply (frule (2) if_live_then_nonz_capD2, simp)
+          apply (clarsimp simp: live_def live_ntfn_def)
+         apply (erule replies_blocked_upd_tcb_st_valid_replies, clarsimp)
+         apply (clarsimp simp: replies_blocked_def st_tcb_at_def obj_at_def)
+        apply (erule fault_tcbs_valid_states_not_fault_tcb_states)
+        apply (clarsimp simp: st_tcb_at_def obj_at_def pred_neg_def)
+       apply (erule delta_sym_refs_remove_only[where tp=TCBSignal], clarsimp)
+        apply (rule subset_antisym, clarsimp)
+        apply (clarsimp simp: state_refs_of_def is_tcb get_refs_def tcb_st_refs_of_def pred_tcb_at_def
+                              obj_at_def)
+        apply (force split: option.splits)
+       apply (rule subset_antisym)
+        apply (clarsimp simp: subset_remove ntfn_q_refs_of_def get_refs_def tcb_st_refs_of_def pred_tcb_at_def
+                              obj_at_def state_refs_of_def)
+        apply (clarsimp split: list.splits option.splits)
+        apply (case_tac list; fastforce)
+       apply (clarsimp simp: subset_remove ntfn_q_refs_of_def get_refs_def tcb_st_refs_of_def pred_tcb_at_def
+                             obj_at_def state_refs_of_def)
+       apply (clarsimp split: list.splits)
+       apply (case_tac list; fastforce)
+      apply (frule (3) not_idle_tcb_in_waitingntfn, fastforce)
+     apply (rule valid_sched_scheduler_act_not_better, clarsimp simp: valid_sched_def)
+     apply (clarsimp simp: st_tcb_at_def obj_at_def pred_neg_def)
+    apply (clarsimp simp: st_tcb_at_def obj_at_def)
+    apply (drule_tac x="hd list" in bspec; clarsimp)+
+  apply (clarsimp simp: invs'_def valid_state'_def valid_pspace'_def valid_tcb_state'_def)
+  apply (intro conjI, clarsimp)
+  apply (clarsimp simp: valid_ntfn'_def)
+  apply (case_tac list; clarsimp simp: valid_ntfn'_def split: list.splits option.splits)
   \<comment> \<open>ActiveNtfn\<close>
-  apply (clarsimp simp add: ntfn_relation_def)
+  apply (clarsimp simp add: ntfn_relation_def Let_def)
   apply (rule corres_guard_imp)
     apply (rule set_ntfn_corres)
-    apply (simp add: ntfn_relation_def combine_ntfn_badges_def
-                     combine_ntfn_msgs_def)
-   apply (simp add: invs_def valid_state_def valid_ntfn_def)
-  apply (simp add: invs'_def valid_state'_def valid_ntfn'_def)
-  done *)
-
-lemma valid_simple'[simp]:
-  "valid_tcb_state' Running = \<top>"
-  "valid_tcb_state' Inactive = \<top>"
-  "valid_tcb_state' Restart = \<top>"
-  "valid_tcb_state' IdleThreadState = \<top>"
-  by (rule ext, simp add: valid_tcb_state'_def)+
+    apply (clarsimp simp: ntfn_relation_def combine_ntfn_badges_def)+
+  done
 
 lemma possibleSwitchTo_ksQ':
-  "\<lbrace>(\<lambda>s. t' \<notin> set (ksReadyQueues s p) \<and> sch_act_not t' s) and K(t' \<noteq> t)\<rbrace>
+  "\<lbrace>(\<lambda>s. t' \<notin> set (ksReadyQueues s p) \<and> sch_act_not t' s) and K (t' \<noteq> t)\<rbrace>
```

Comment:
nitpick: since you're already touching this line
```suggestion
  "\<lbrace>\<lambda>s. t' \<notin> set (ksReadyQueues s p) \<and> sch_act_not t' s \<and> t' \<noteq> t\<rbrace>
```